core/chains/evm/logpoller/log_poller_internal_test.go

```diff
@@ -346,54 +346,68 @@ func TestLogPoller_Replay(t *testing.T) {
 		}()
 		assert.ErrorIs(t, lp.Replay(ctx, 4), ErrReplayInProgress)
 		<-done
+		lp.replayComplete <- nil
+		lp.wg.Wait()
 	})
 
 	// Main lp.run() loop shouldn't get stuck if client aborts
 	t.Run("client abort doesnt hang run loop", func(t *testing.T) {
 		lp.backupPollerNextBlock = 0
 
-		timeout := time.After(1 * time.Second)
-		lp.ctx, lp.cancel = context.WithCancel(tctx)
+		timeout := time.After(2 * time.Second)
 		ctx, cancel := context.WithCancel(tctx)
 
 		var wg sync.WaitGroup
 		wg.Add(2)
-		ec.On("FilterLogs", lp.ctx, mock.Anything).Once().Return([]types.Log{log1}, nil).Run(func(args mock.Arguments) {
+
+		ec.On("FilterLogs", mock.Anything, mock.Anything).Once().Return([]types.Log{log1}, nil).Run(func(args mock.Arguments) {
 			go func() {
 				defer wg.Done()
 				assert.ErrorIs(t, lp.Replay(ctx, 4), ErrReplayInProgress)
 			}()
 		})
-		ec.On("FilterLogs", lp.ctx, mock.Anything).Once().Return([]types.Log{log1}, nil).Run(func(args mock.Arguments) {
+		ec.On("FilterLogs", mock.Anything, mock.Anything).Once().Return([]types.Log{log1}, nil).Run(func(args mock.Arguments) {
 			go func() {
 				defer wg.Done()
 				cancel()
 				lp.replayStart <- 4
 			}()
 		})
-		ec.On("FilterLogs", lp.ctx, mock.Anything).Return([]types.Log{log1}, nil).Maybe()
+
+		ec.On("FilterLogs", mock.Anything, mock.Anything).Return([]types.Log{log1}, nil).Maybe() // in case task gets delayed by >= 100ms
+
+		lp.ctx, lp.cancel = context.WithCancel(tctx)
 		lp.wg.Add(1)
-		done := make(chan struct{})
+		defer func() {
+			select {
+			case <-lp.replayStart:
+			default:
+			}
+			wg.Wait()
+			lp.cancel()
+			lp.wg.Wait()
+		}()
+
 		go func() {
-			defer close(done)
 			lp.run()
 		}()
 		select {
 		case <-timeout:
-			assert.Fail(t, "lp.run() got stuck--failed to respond to second replay event within 1s")
+			assert.Fail(t, "lp.run() got stuck--failed to respond to second replay event within 2s")
 		case <-utils.WaitGroupChan(&wg):
-			lp.cancel()
 		}
-		<-done
 	})
 
-	lp.wg.Wait() // ensure logpoller has exited before continuing on to next test
+	// remove Maybe expectation from prior subtest, as it will override all expected calls in future subtests
+	ec.On("FilterLogs", mock.Anything, mock.Anything).Return([]types.Log{log1}, nil).Maybe().Unset()
```

Comment:
I don't understand this line. `Maybe` already passes if something is not called, so why would we need to remove it? And isn't this line both adding and removing at once? i.e. Wouldn't we need a reference to the previous test's `Maybe()` call to be able to unset that instance?

Reply:
If you remove this line, then the next test fails because the `Maybe()` expectation takes precedence over the more specific expectations it adds to the mock:
```
		ec.On("FilterLogs", mock.Anything, mock.Anything).Once().Return([]types.Log{log1}, nil).Run(func(args mock.Arguments) {
			go func() {
				defer wg.Done()
				lp.replayStart <- 4
			}()
		})
		ec.On("FilterLogs", mock.Anything, mock.Anything).Once().Return([]types.Log{log1}, nil).Run(func(args mock.Arguments) {
			defer wg.Done()
			lp.cancel()
		})
```
As I understand it, when you add multiple `Once()` expectations for the same method call with same arguments, the `Called()` method knows that they need to happen in order one at a time.  As soon as a `Maybe()` gets added to this ordered list of expectations, nothing that comes after that has any effect because the `Maybe` will match any number of calls that happen after that.

My previous fix for this was to use `lp.ctx` instead of `mock.Anything`, since the context is different between the two tests that makes them each have completely separate call expectations.  But this leads to the race condition you found.

The "adding and removing" is a good point... it might be less confusing if instead we just save the previous call in some test-scoped variable (`c = ec.On(...).Maybe()`), and then call `c.Unset()` after the test exits?

In the docstring, they do something more similar to what I've done (although I probably didn't need the `.Maybe()`:
```
// Unset removes a mock handler from being called.
//    test.On("func", mock.Anything).Unset()
```
It will search through the whole list of expectations and remove anything that matches the method call and argument list.  So it doesn't actually matter how many copies of the expectation are added before the `Unset()`, it will remove them all.

Reply:
Yes, it would be more clear w/o the `Maybe`, if that is equivalent.

Reply:
Also, w/o the return:
```suggestion
	ec.On("FilterLogs", mock.Anything, mock.Anything).Unset()
```